### PR TITLE
✨ feat: deterministic nonce derivation

### PR DESCRIPTION
**Pull request by zpv**
## What

Deterministic nonce derivation is added so in the case of dataloss, given the same `eventID`, the nonces can be recreated.
These nonces are derived from the index for a given outcome plus an added tweak based off a taggedHash of `eventID`

```typescript

  public async getPrivkeyForIndex(index: number) {
    const derivationPath = `${R_VAL_PURPOSE}'/${this.network.coinType}'/0'/0/${index}`;
    const keyPair = await this.client.getMethod("keyPair")(derivationPath);

    return keyPair.privateKey;
  }

  private calcNonceTweak(eventId: string, privateKey: Buffer): Buffer {
    const nonceTweak = CfdUtils.GetSchnorrPubkeyFromPrivkey(privateKey);
    const bytes = Buffer.concat([nonceTweak, Buffer.from(eventId, "hex")]);

    return getTaggedHash("DLC/oracle/nonce/v0", bytes);
  }

  private async getKValue(eventId: string, index: number) {
    const privateKey = await this.getPrivkeyForIndex(index);
    const nonceTweak = this.calcNonceTweak(eventId, privateKey);

    return CfdUtils.TweakAddPrivkey(this.privateKey, nonceTweak);
  }
```

### CODE REVIEW

**Comment by matthewjablack** [comment]
Note: we need to check if https://github.com/discreetlogcontracts/dlcspecs/pull/167 affects anything in our implementation

**Review by dillionverma — APPROVED** [review]
Couple small comments and questions but otherwise looks good to me! 🔥

File: src/database/announcements/announcement.types.ts [hunk 1]
```diff
@@ -2,10 +2,9 @@ import { Document, Model } from "mongoose";
 
 export interface IAnnouncement {
   offerId: string;
-  oracleName: string;
   uri: string;
   announcement: string;
-  attestment: string;
+  attestment?: string;
   cso: boolean;
   eventId: string;
```
Comment (matthewjablack) [inline src/database/announcements/announcement.types.ts:9]: We should be ensuring the uniqueness of `eventId`. Will this be done in subsequent PR?

File: src/api/index.ts [hunk 2]
```diff
@@ -25,24 +19,6 @@ import {
...
-async function getNodePubKey(
-  client: FinanceClient,
-  network: BitcoinNetwork
-): Promise<Buffer> {
-  const derivationPath = `${MessageType.DlcOfferV0}'/${network.coinType}'/0'/0/0`;
-  const keyPair = await client.getMethod("keyPair")(derivationPath);
-  return Buffer.from(keyPair.publicKey, "hex");
-}
-
 const postAttest = async (
```
Comment (matthewjablack) [inline src/api/index.ts:22]: I'm guessing we'll remove this from API in subsequent PR as well

File: src/database/announcements/announcement.types.ts [hunk 3]
```diff
@@ -2,10 +2,9 @@ import { Document, Model } from "mongoose";
 
 export interface IAnnouncement {
```
Comment (matthewjablack) [inline src/database/announcements/announcement.types.ts:3]: We also need an additional model `Nonce` where we can ensure the uniqueness of each Nonce in the DB as discussed

File: src/lib/Oracle.ts [hunk 4]
```diff
@@ -1,73 +1,161 @@
...
-    const signRequest: cfdjs.SchnorrSignRequest = {
-      privkey: this.privateKey,
-      message,
-      nonceOrAux: this.kValues[significantDigit - 1],
-      isNonce: true,
-      isHashed: true,
-    };
-
-    return cfdjs.SchnorrSign(signRequest).hex;
+  public async createAttestation(oracleEvent: OracleEventV0, outcome: number) {
```
Comment (matthewjablack) [inline src/lib/Oracle.ts:112]: Would be good to run validation on `oracleEvent` (`oracleEvent.validate()`) to make sure we validate the basic fields

File: src/lib/Oracle.ts [hunk 5]
```diff
@@ -1,73 +1,161 @@
...
+    const privateKey = await this.getPrivkeyForIndex(index);
+    const nonceTweak = this.calcNonceTweak(eventId, privateKey);
+
+    return CfdUtils.TweakAddPrivkey(this.privateKey, nonceTweak);
+  }
+
+  public async createNewOracleEventAnnouncement(
+    eventId: string,
+    eventMaturityEpoch: number,
+    eventDescriptor: DigitDecompositionEventDescriptorV0
```
Comment (matthewjablack) [inline src/lib/Oracle.ts:65]: Might be good to run `eventDescriptor.validate()`

File: src/lib/Oracle.ts [hunk 6]
```diff
@@ -1,73 +1,161 @@
...
+  ) {
+    if (await AnnouncementModel.exists({ eventId })) {
+      throw new Error(
+        `Oracle announcement for event ${eventId} already exists`
+      );
+    }
+
+    const event = new OracleEventV0();
+    const noncesNeeded =
+      eventDescriptor.nbDigits + (eventDescriptor.isSigned ? 1 : 0);
```
Comment (matthewjablack) [inline src/lib/Oracle.ts:75]: node-dlc doesn't currently support `isSigned` but this is good for future proofing https://github.com/AtomicFinance/node-dlc/blob/master/packages/messaging/lib/messages/EventDescriptor.ts#L171

**Review by matthewjablack — COMMENTED** [review]
A couple of comments for improvement

Overall looks solid 🔥 Thanks for paying particular attention to security on this one

File: src/api/index.ts [hunk 7]
```diff
@@ -70,45 +44,24 @@ const postAttest = async (
...
-  //     .taggedHash("DLC/oracle/attestation/v0", outcomes[i].toString())
-  //     .toString("hex");
-  //   sigs.push(Buffer.from(oracle.GetSignature(m, i + 1), "hex"));
-  // }
-
-  const attestation = new OracleAttestationV0();
-  attestation.eventId = announcement.oracleEvent.eventId;
-  attestation.signatures = sigs;
-  attestation.outcomes = outcomes;
+  const oracle = await Oracle.BuildOracle(network, client);
```
[the remark this answers is not in the document]
Comment (zpv) [reply src/api/index.ts:60]: Separate PR for now. API structure will be refactored when tackling the endpoints tickets.

**Pull request merged**